bookshelf/test/lib/DataModelTestTrait.php

```diff
@@ -0,0 +1,86 @@
+<?php
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+namespace Google\Cloud\Samples\Bookshelf;
+
+/**
+ * Common code to test all data models.
+ *
+ * @property DataModelInterface $model Must be set by parent class.
+ */
+trait DataModelTestTrait
+{
+    /**
+     * Tests all the crud operations in a data model.
+     * It iterates through all the books in the data model, so for performance
+     * reasons, the data model should not be prefilled with hundreds of books.
+     */
+    function testDataModel()
+    {
+        $model = $this->model;
+        // Iterate over the existing books and count the rows.
+        $fetch = array('cursor' => null);
+        $rowCount = 0;
+        do {
+            $fetch = $model->listBooks(10, $fetch['cursor']);
+            $rowCount += count($fetch['books']);
+        } while ($fetch['cursor']);
+
+        // Insert two books.
+        $breakfastId = $model->create(array(
+            'title' => 'Breakfast of Champions',
+            'author' => 'Kurt Vonnegut'
+        ));
+
+        $bellId = $model->create(array(
+            'title' => 'For Whom the Bell Tolls',
+            'author' => 'Ernest Hemingway'
+        ));
+
+        // Iterate over the books again and verify there are now 2 more.
+        $newCount = 0;
+        do {
+            // Only fetch one book at a time to test that code path.
+            $fetch = $model->listBooks(1, $fetch['cursor']);
+            $newCount += count($fetch['books']);
+        } while ($fetch['cursor']);
+        $this->assertEquals($rowCount + 2, $newCount);
+
+        // Make sure the book we read looks like the book we wrote.
+        $breakfastBook = $model->read($breakfastId);
+        $this->assertEquals('Breakfast of Champions', $breakfastBook['title']);
+        $this->assertEquals('Kurt Vonnegut', $breakfastBook['author']);
+        $this->assertEquals($breakfastId, $breakfastBook['id']);
+        $this->assertNull($breakfastBook['description']);
+
+        // Try updating a book.
+        $breakfastBook['description'] = 'A really fun read.';
+        $model->update($breakfastBook);
+        $breakfastBookCopy = $model->read($breakfastId);
+        // And confirm it was correctly updated.
+        $this->assertEquals(
+            'A really fun read.',
+            $breakfastBookCopy['description']
+        );
```

Comment:
Very nice and extensive test suites all around! Some suggestions for this test:

 - test adding a property that doesn't exist
 - test updating a property that doesn't exist
 - test updating a single property and leaving the rest out, verify the other properties are not *empty* after saving
 - test updating a property and setting it empty, and make sure it *is* empty after you save

This would require splitting this into multiple functions